connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java

```diff
@@ -0,0 +1,227 @@
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
+        MetricGroup group = groupsByName.get(groupName);
+        if (group == null) {
+            Map<String, String> tags = tags(includeWorkerId ? workerId : null, tagKeyValues);
+            group = new MetricGroup(groupName, tags);
+            MetricGroup previous = groupsByName.putIfAbsent(groupName, group);
+            if (previous != null) group = previous;
+        }
+        return group;
+    }
+
+    /**
+     * Get the time.
+     *
+     * @return the time; never null
+     */
+    public Time time() {
+        return time;
+    }
+
+    /**
+     * Stop and unregister the metrics from any reporters.
+     */
+    public void stop() {
+        metrics.close();
+        LOG.debug("Unregistering Connect metrics with JMX for worker '{}'", workerId);
+        AppInfoParser.unregisterAppInfo(JMX_PREFIX, workerId);
+    }
+
+    /**
+     * A group of metrics. Each group maps to a JMX MBean and each metric maps to an MBean attribute.
+     */
+    public class MetricGroup {
+        private final String groupName;
+        private final Map<String, String> tags;
+
+        protected MetricGroup(String groupName, Map<String, String> tags) {
+            this.groupName = groupName;
+            this.tags = tags;
+        }
+
+        /**
+         * Create the name of a metric that belongs to this group and has the group's tags.
+         *
+         * @param name the name of the metric/attribute; may not be null
+         * @param desc the description for the metric/attribute; may not be null
+         * @return the metric name; never null
+         */
+        public MetricName metricName(String name, String desc) {
+            return metrics.metricName(name, groupName, desc, tags);
+        }
+
+        /**
+         * The {@link Metrics} that this group belongs to.
+         *
+         * @return the metrics; never null
+         */
+        public Metrics metrics() {
+            return metrics;
+        }
+
+        Map<String, String> tags() {
+            return Collections.unmodifiableMap(tags);
+        }
+    }
+
+    /**
+     * Create a set of tags using the supplied key and value pairs.
+     *
+     * @param workerId the worker ID that should be included first in the tags; may be null if not to be included
+     * @param keyValue the key and value pairs for the tags; must be an even number
+     * @return the map of tags that can be supplied to the {@link Metrics} methods; never null
+     */
+    static Map<String, String> tags(String workerId, String... keyValue) {
+        if ((keyValue.length % 2) != 0)
+            throw new IllegalArgumentException("keyValue needs to be specified in pairs");
+        Map<String, String> tags = new HashMap<>();
+        if (workerId != null && !workerId.trim().isEmpty()) {
```

Comment:
might want to rename `workerId` so it doesn't shadow the member field. something like `workerIdOpt` could work

Reply:
ack

Reply:
ah, right. nah, that's fine. just when reviewing I had the thought that if we guaranteed non-`null`/non-empty in the constructor, this wouldn't be necessary. i realized that it was actually intentional, but easy to miss when reviewing here and not getting the same highlighting as an IDE